Feature/fxls8974 driver

CZKikin opened the pull request:
Added drivers for the fxls8974 accelerometer, tested over I2C on the mimxrt1040_evk board.

CZKikin commented:
While fixing bugs I found TODO comment in samples/sensor/sensor_shell/sample.yml which should've been deleted since the https://github.com/zephyrproject-rtos/zephyr/issues/63414 is resolved.

manuargue commented on drivers/sensor/nxp/fxls8974/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# Makefile - FXLS8974 3-axis accelerometer
+#
+# Copyright (c) 2024, NXP Semiconductor
```
```suggestion
# Copyright 2024 NXP
```

manuargue commented on drivers/sensor/nxp/fxls8974/Kconfig:
```diff
@@ -0,0 +1,58 @@
+# FXLS8974 3-axis accelerometer
+
+# Copyright (c) 2024 Freescale Semiconductor, Inc.
```
```suggestion
# Copyright 2024 NXP
```

manuargue commented on drivers/sensor/nxp/fxls8974/fxls8974.c:
```diff
@@ -0,0 +1,629 @@
+/*
+ * Copyright (c) 2024 NXP Semiconductor
```
```suggestion
 * Copyright 2024 NXP
```

manuargue commented on drivers/sensor/nxp/fxls8974/fxls8974.h:
```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2024 NXP Semiconductor
```
```suggestion
 * Copyright 2024 NXP
```

manuargue commented on drivers/sensor/nxp/fxls8974/fxls8974_trigger.c:
```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2024 NXP Semiconductor
```
```suggestion
 * Copyright 2024 NXP
```

manuargue commented on dts/bindings/sensor/nxp,fxls8974-common.yaml:
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2024, NXP
```
```suggestion
# Copyright 2024 NXP
```

manuargue commented on dts/bindings/sensor/nxp,fxls8974-i2c.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024, NXP
```
```suggestion
# Copyright 2024 NXP
```

manuargue commented on dts/bindings/sensor/nxp,fxls8974-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024, NXP
```
```suggestion
# Copyright 2024 NXP
```

manuargue commented on samples/sensor/fxls8974/src/main.c:
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2024 Nxp Semiconductor
```
```suggestion
 * Copyright 2024 NXP
```

manuargue reviewed (dismissed):
nits on NXP copyright notice to comply with OSS company policies

manuargue commented:
> [doc: updated copyrights](https://github.com/zephyrproject-rtos/zephyr/pull/76071/commits/e138863329675725da9a6fff17b5415c36291e55)

@CZKikin please check https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow, the workflow in Zephyr is that you need to interactively rebase commits to fix review issues. Please make sure you are following the process

hakehuang commented on samples/sensor/fxls8974/sample.yaml:
```diff
@@ -0,0 +1,26 @@
+tests:
+  sample.sensor.fxls8974.accel:
+    platform_allow:
+      - mimxrt1040_evk
```
we can use 
`filter: dt_compat_enabled("nxp,fxls8974")`

hakehuang commented on samples/sensor/fxls8974/sample.yaml:
```diff
@@ -0,0 +1,25 @@
+      - "^AX=\\W*[0]\\..*AY=\\W*[0]\\..*AZ=\\W*([9]|[1])([0-1]|\\.).*$"
+      - "^AX=\\W*[0]\\..*AY=\\W*[0]\\..*AZ=\\W*([9]|[1])([0-1]|\\.).*$"
+      - "^AX=\\W*[0]\\..*AY=\\W*[0]\\..*AZ=\\W*([9]|[1])([0-1]|\\.).*$"
+      - "^AX=\\W*[0]\\..*AY=\\W*[0]\\..*AZ=\\W*([9]|[1])([0-1]|\\.).*$"
```
![image](https://github.com/user-attachments/assets/a462a50d-42df-41ca-9290-0a97f05ef658)
the regex match is wrong

CZKikin replied:
![obrazek](https://github.com/user-attachments/assets/06895dc9-6b27-4c35-af9b-0cbecff9c965)
Seems to work for me

hakehuang replied:
the regex match shall compatible to all the situations, stick to one situation is not reasonable.

CZKikin replied:
I've added a vector cross to the example, so the regex won't have to be stupid long, tested on 1064, test passes in all orientations.

hakehuang commented on samples/sensor/fxls8974/sample.yaml:
```diff
@@ -0,0 +1,26 @@
+    extra_configs:
+      - CONFIG_FXLS8974_TRIGGER_OWN_THREAD=y
+      - CONFIG_FXLS8974_DRDY_INT1=y
+    skip: true # remove after a board with connected interrupt is added
```
is this need, if the filter is there?

CZKikin replied:
Yes, some boards do not have interrupt pins connected to the accelerometer.

hakehuang replied:
then please use platform_exclude, with this setting all boards will be skipped.

MaureenHelm commented on dts/bindings/sensor/nxp,fxls8974-common.yaml:
```diff
@@ -0,0 +1,41 @@
+      the signal that is presented to the driver.
+
+  range:
+    type: int
+    default: 16
+    description: Range in g
+    enum:
+      - 16 # 16g (7.81 mg/LSB)
+      - 8 # 8g (3.91 mg/LSB)
+      - 4 # 4g (1.95 mg/LSB)
+      - 2 # 2g (0.98 mg/LSB)
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

Also need to move the comments on range encodings into the description so they'll be rendered in the documentation.

CZKikin replied:
We've added the sample because we wanted to have a way to test out the accelerometer driver with interrupts, similar to the fxos8600 sample. Also there may be more accelerometers from NXP which could run on this driver in the future. Is there any sample which tests accelerometers with interrupts?

MaureenHelm reviewed (changes requested):
Please use `samples/sensor/accel_polling/` instead of introducing a new driver-specific sample.

hakehuang reviewed (dismissed):
board testing pass on mimxrt1040_evk now, and @CZKikin please update according Maureen's comments.

Commit:
drivers: sensor: fxls8974: Added driver for this accelerometer

Added driver for the fxls8974 accelerometer which is present on the
mimxrt1040_evk board.

Signed-off-by: Krystof Sadlik <krystof.sadlik@nxp.com>

Commit:
boards: mimxrt1040_evk: Added pinctrl for fxls8974

Added fxls8974 pinctrl to the dts file

Signed-off-by: Krystof Sadlik <krystof.sadlik@nxp.com>

Commit:
dts: bindings: Added bindings for the fxls8974 accelerometer

Added bindings to enable fxls8974 accelerometer

Signed-off-by: Krystof Sadlik <krystof.sadlik@nxp.com>

Commit:
Samples: Sensor: Added mimxrt1040 config, and fxls8974

Added mimxrt1040_evk to inregration platforms, so that fxls8974 is tested,
and added config for mimxrt1040_evk to sensor_shell

Signed-off-by: Krystof Sadlik <krystof.sadlik@nxp.com>

MaureenHelm commented on drivers/sensor/nxp/fxls8974/fxls8974.c:
```diff
@@ -0,0 +1,632 @@
+		uint8_t old_val;
+
+		if (fxls8974_byte_read_spi(dev, reg, &old_val) < 0)
+			return -EIO;
```
Add braces to every `if`, `else`, `do`, `while`, `for` and `switch` body, even for single-line code blocks
https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

MaureenHelm commented on drivers/sensor/nxp/fxls8974/fxls8974.c:
```diff
@@ -0,0 +1,632 @@
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+		if (cfg->inst_on_bus == FXLS8974_BUS_I2C) {
+			if (!device_is_ready(cfg->bus_cfg.i2c.bus)) {
```
Use `i2c_is_ready_dt`

MaureenHelm commented on drivers/sensor/nxp/fxls8974/fxls8974.c:
```diff
@@ -0,0 +1,632 @@
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+		if (cfg->inst_on_bus == FXLS8974_BUS_SPI) {
+			if (!device_is_ready(cfg->bus_cfg.spi.bus)) {
```
Use `spi_is_ready_dt`

MaureenHelm reviewed (changes requested):
Sorry I didn't catch this before, but the new sensor also needs to be added to `tests/drivers/build_all/sensor`.

Otherwise LGTM

Commit:
tests: build_all: sensors: Added fxls8974

Added fxls8974 to build_all tests

Signed-off-by: Krystof Sadlik <krystof.sadlik@nxp.com>

hakehuang reviewed (approved):
board testing pass
